agent/fw_drupal8.c

```diff
@@ -605,6 +605,94 @@ NR_PHP_WRAPPER(nr_drupal94_invoke_all_with_clean) {
 NR_PHP_WRAPPER_END
 #endif  // OAPI
 
+static bool nr_validate_key_val_arr(nr_php_string_hash_key_t* key, zval* val) {
+  if (NULL == key || 0 == ZEND_STRING_LEN(key)
+      || 0 == nr_php_is_zval_valid_array(val)
+      || 0 == zend_hash_num_elements(Z_ARRVAL_P(val))) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+/*
+ * Purpose: Instrument Drupal Attribute Hooks for Drupal 11.1+
+ *
+ * Params: 1. A zval pointer to the moduleHandler instance in use by Drupal.
+ *
+ * Return: bool
+ *
+ */
+static bool nr_drupal_hook_attribute_instrument(zval* module_handler) {
+  zval* hook_implementation_map = NULL;
+
+  nr_php_string_hash_key_t* hook_key = NULL;
+  zval* hook_val = NULL;
+  nr_php_string_hash_key_t* class_key = NULL;
+  zval* class_val = NULL;
+  nr_php_string_hash_key_t* method_key = NULL;
+  zval* module_val = NULL;
+
+  char* hookpath = NULL;
+
+  hook_implementation_map = nr_php_get_zval_object_property(
+      module_handler, "hookImplementationsMap");
+
+  if (!nr_php_is_zval_valid_array(hook_implementation_map)) {
+    nrl_warning(NRL_FRAMEWORK,
+                "hookImplementationsMap property not a valid array");
+    return false;
+  }
+
+  ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL_P(hook_implementation_map), hook_key,
+                                hook_val) {
+    if (nr_validate_key_val_arr(hook_key, hook_val)) {
+      nrl_warning(NRL_FRAMEWORK,
+                  "hookImplementationsMap[hook]: invalid key or value");
+      return false;
+    }
+
+    ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL_P(hook_val), class_key, class_val) {
+      if (nr_validate_key_val_arr(class_key, class_val)) {
+        nrl_warning(NRL_FRAMEWORK,
+                    "hookImplementationsMap[class]: invalid key or value");
+        return false;
+      }
+
+      ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL_P(class_val), method_key,
+                                    module_val) {
+        if (NULL == method_key
+            || 0 == nr_php_is_zval_valid_string(module_val)) {
+          nrl_warning(NRL_FRAMEWORK,
+                      "hookImplementationsMap[method]: invalid key or value");
+          return false;
+        }
+
+        if (0
+            == nr_stricmp(ZEND_STRING_VALUE(class_key),
+                          "Drupal\\Core\\Extension\\ProceduralCall")) {
+          hookpath = nr_formatf("%s", ZEND_STRING_VALUE(method_key));
```

Comment:
What is more likely:
- `class_key` starting with `Drupal\Core`
- `class_key` ending with `Drupal\Core\Extension\ProceduralCall`?

Or put it in a different way: how often will `nr_stricmp` have to scan through `Drupal\Core` only to fail on the next character? Is it possible to optimize this check so that it fails sooner than later?

Reply:
I wonder if using `nr_striendswith(ZEND_STRING_VALUE(class_key), ZEND_STRING_LEN(class_key), NR_PSTR("Drupal\\Core\\Extension\\ProceduralCall"))` wouldn’t be more optimal. Worst case, `nr_striendswith` will still have to scan the whole string. Best case, `nr_striendswith` will fail fast on length comparison.

Reply:
I agree with @lavarou 's suggestion; that's a good optimization.

Reply:
315ecbb1b8d7e0687d239eaa84b03eb8816ccbcf